framework/io/gmsh_io.cc

```diff
@@ -0,0 +1,412 @@
+  file.seekg(0);
+  while (std::getline(file, file_line))
+    if (node_section_name == file_line)
+      break;
+
+  std::getline(file, file_line);
+  iss = std::istringstream(file_line);
+  int num_nodes;
+  if (not(iss >> num_nodes))
+    throw std::logic_error(fname + ": Failed while trying to read "
+                                   "the number of nodes.");
+
+  auto& vertices = mesh->Vertices();
+  vertices.clear();
+  vertices.resize(num_nodes);
+
+  for (int n = 0; n < num_nodes; n++)
+  {
+    std::getline(file, file_line);
+    iss = std::istringstream(file_line);
+
+    int vert_index;
+    if (not(iss >> vert_index))
+      throw std::logic_error(fname + ": Failed to read vertex index.");
+
+    if (not(iss >> vertices[vert_index - 1].x >> vertices[vert_index - 1].y >>
+            vertices[vert_index - 1].z))
+      throw std::logic_error(fname + ": Failed while reading the vertex "
+                                     "coordinates.");
+  }
+
+  // Define utility lambdas
+  /**Lambda for reading nodes.*/
+  auto ReadNodes = [&iss, &fname](int num_nodes)
+  {
+    std::vector<int> raw_nodes(num_nodes, 0);
+    for (int i = 0; i < num_nodes; ++i)
+      if (not(iss >> raw_nodes[i]))
+        throw std::logic_error(fname + ": Failed when reading element "
+                                       "node index.");
+
+    std::vector<uint64_t> nodes(num_nodes, 0);
+    for (int i = 0; i < num_nodes; ++i)
+      if ((raw_nodes[i] - 1) >= 0)
+        nodes[i] = raw_nodes[i] - 1;
+    return nodes;
+  };
+
+  /**Lamda for checking if an element is 1D.*/
+  auto IsElementType1D = [](int element_type)
+  {
+    if (element_type == 1)
+      return true;
+
+    return false;
+  };
+
+  /**Lamda for checking if an element is 2D.*/
+  auto IsElementType2D = [](int element_type)
+  {
+    if (element_type == 2 or element_type == 3)
+      return true;
+
+    return false;
+  };
+
+  /**Lamda for checking if an element is 2D.*/
+  auto IsElementType3D = [](int element_type)
+  {
+    if (element_type >= 4 and element_type <= 7)
+      return true;
+
+    return false;
+  };
+
+  /**Lambda for checking supported elements.*/
+  auto IsElementSupported = [](int element_type)
+  {
+    if (element_type >= 1 and element_type <= 7)
+      return true;
+
+    return false;
+  };
+
+  /**Lambda giving the cell subtype, given the MSH cell type.*/
+  auto CellTypeFromMSHTypeID = [](int element_type)
+  {
+    CellType cell_type = CellType::GHOST;
+
+    if (element_type == 1)
+      cell_type = CellType::SLAB;
+    else if (element_type == 2)
+      cell_type = CellType::TRIANGLE;
+    else if (element_type == 3)
+      cell_type = CellType::QUADRILATERAL;
+    else if (element_type == 4)
+      cell_type = CellType::TETRAHEDRON;
+    else if (element_type == 5)
+      cell_type = CellType::HEXAHEDRON;
+    else if (element_type == 6 or // Prism
+             element_type == 7)
+      cell_type = CellType::POLYHEDRON; // Pyramid
```

Comment:
should // Pyramid but moved up one line to be next to type 7 ?

Reply:
Strictly speaking those magic numbers should not be there at all. They should be an enum. But, I only moved the code. This routine will be replaced by [gmshparsercpp](https://github.com/andrsd/gmshparsercpp) which does the right thing with the enum, see https://github.com/andrsd/gmshparsercpp/blob/main/include/gmshparsercpp/Enums.h#L11